### DISCUSSION TITLE
Development Blog

### DISCUSSION THREAD

**Discussion by nefarius** [CircumSpector/DS4Windows#21] [General]
As we're plowing through the code and change stuff I think it's helpful to keep "the public" up to date what's happening, besides the wave of commit messages, so I *try* to start a development diary here in a semi-regular fashion.

Wish me luck since you know, I'm the master of inconsistency 🤣 

Let's go!

**Comment by nefarius** [comment]
# 2021-12-22

## What's happened so far

I've started this blog way after the rework has started so this post will be used to recall and describe roughly what has happened under the hood *so far*.

### Switched from using XML to JSON

Everything that needs to be persisted (application configuration, controller properties, and last but not least the profiles) will be using JSON instead of XML. Reason being that the entire project doesn't really use anything that warrants use of a very verbose and complex markup language like XML, it's easier to read for the human eye (and to edit) and battle tested serializers in .NET exist for it in ages.

This does *not* mean that every customization previous DS4Windows users made (profiles, macros, etc.) is lost, there are XML to JSON converters in development, that will take care to transparently migrate >90% of the existing configuration, where possible.

A ton of hours was dumped into dissecting and replacing the existing XML (de-)serializer **which was hand-crafted!** It was a monster; tons of duplicated and temporary variables, string and indexes used everywhere to keep track of states and objects... Nobody will miss this beast 😁 The way data gets persisted also has been completely redesigned and follows a true object-oriented pattern now. A "profile" is now a POCO with an object graph that is (de-)serializable with ease and is very simple to extend with features in the future.

### UI rework

[Adonis UI](https://benruehl.github.io/adonis-ui/) was introduced to unify the look and feel and also make the styling easier. This may or may not be the permanent solution as there are many other frameworks out there that offer more features but for now it's enough to give the UI a fresh look and also a more coherent dark theme.

![image](https://user-images.githubusercontent.com/286631/147143414-08f58953-5cc0-44d7-abf5-7e6be699fd37.png)

![image](https://user-images.githubusercontent.com/286631/147143443-4fb5a20e-d57b-458e-a93f-4926129bff6d.png)

![image](https://user-images.githubusercontent.com/286631/147143471-bdb52fd8-07e0-4827-9188-d2e1a5804d37.png)

![image](https://user-images.githubusercontent.com/286631/147143478-ed5b828d-4d00-44f5-9c09-1c94ca1c0c6e.png)

![image](https://user-images.githubusercontent.com/286631/147143506-fc297810-2bf9-4410-8546-45719bfb8851.png)

![image](https://user-images.githubusercontent.com/286631/147143521-1f0fbf5b-108b-43e4-855b-f952d0b7a095.png)

The changes may be subtle but it makes working with XAML and WPF in general more fun 😀 

### Rewriting Profile Editor

This is a huge but important task. The entire profiling system is based upon design sin after design sin which makes it incredibly hard to understand, or even extend. Instead of having an object-oriented approach to store, retrieve and organise profile data countless arrays, indexes, fields and copies of fields, references here and there were used. This mess needs to be untangled and redesigned.

This is currently ongoing full throttle.

### Addressing and fixing many memory leaks

Disclaimer: technically it's not really leaking memory since in a managed language like C# it's quite hard (albeit not impossible) to actually allocate memory that is never freed again during application runtime, but for the sake of simplicity lets stick to the common term when a lot of heap memory is allocated that puts a lot of pressure on the garbage collector and leaves big lumps of unused memory dangling around for many seconds, if not minutes.

With that out of the way the input handling logic (reading packets from the actual hardware and transforming it to be fed to the virtual output devices) has been optimised in a major way. Previously heap memory usage was running away slowly but steadily during program usage due to many unnecessary allocations and improper use of the HID-Library abstraction layer. With these fixes in place, memory usage during runtime with one or more controllers connected was greatly improved and stabilized.

A lot of design sins also hide in the Profile Editor implementation, the classes involved are currently ported over to utilize IoC (dependency injection) so singletons will remain in memory once and just the relevant references will be updated during runtime instead of allocating and destroying a ton of objects when invoking the Profile Editor and related types.

**Comment by nefarius** [comment]
# 2021-12-27 🎅 

Currently reworking the Special Actions types so they can be (de-)serialized properly. Then I can migrate the Special Actions View Model to IoC and a huge chunk of the Profile Editor will be running under new logic!

**Comment by nefarius** [comment]
# 2021-12-29

Wow, even though I wrecked havoc on the entire profile editing pipeline, the default behaviour and remapping to a virtual X360 controller **still works**!

![image](https://user-images.githubusercontent.com/286631/147663298-b0f80d1b-2750-4be5-85e1-b2fa1b004273.png)

![image](https://user-images.githubusercontent.com/286631/147663306-72381969-16cb-4f2c-9d43-7b98b201f43b.png)

Mainly because I have not yet ripped away all the global fields, properties and methods that the core services depends on but still, wow. This decoupling shows how much redundant code there is, the old XML-style profiles are still loaded and read all over the place, I am actually impressed!

**Comment by nefarius** [comment]
# 2022-01-01

Added a small new feature; allowing to ship immutable (non-user-editable) profiles to provide some useful presets that can not be altered on accident ("Default" in this example screenshot is immutable):

![image](https://user-images.githubusercontent.com/286631/147842048-3e79e5f5-1aff-4e74-9d8a-039178e1478c.png)

**Comment by nefarius** [comment]
# 2022-01-16

Currently reworking the tray area logic. Migrated profile enumeration to new profile service, removed a lot of redundant events, applied consistent dark theme:

![image](https://user-images.githubusercontent.com/286631/149646594-406f6b58-7ac5-4962-b02a-66c85012c328.png)

Added experimental tooltip bubble styling, will rework this to be bigger, easier to read, consistent with theme and simpler to populate:

![image](https://user-images.githubusercontent.com/286631/149646633-d670e638-2a04-46ed-88fb-faa48705acab.png)

Lots of smaller ongoing changes of clean-up and rewrites under the hood.

**Reply by nefarius** [reply]
Preview of a few new features (clarification dialogs to address common issues):

## Wrong architecture

![image](https://user-images.githubusercontent.com/286631/149647140-8bc2029d-d811-411b-8a01-d43bf74f5ac2.png)

## Conflict with Steam

![image](https://user-images.githubusercontent.com/286631/149647149-351954ec-9e3f-4171-92eb-8ae3c9a13366.png)

## Tracing is enabled

![image](https://user-images.githubusercontent.com/286631/149647168-eba470c0-77c3-455b-8446-b3046634f297.png)

**Reply by nefarius** [reply]
## Windows 11 Bashing 😁 

...because it deserves it.

![image](https://user-images.githubusercontent.com/286631/149648555-70776078-d06d-47c6-849b-6d9572fc75a0.png)

**Reply by nefarius** [reply]
Memory allocation happening when you simply drag the window around 😲 

![image](https://user-images.githubusercontent.com/286631/149648808-a8370cc1-3380-4718-a131-3f7afa546d12.png)

**Reply by nefarius** [reply]
I am porting over the USB/HID device (un)plug detection to individual modules/services/extensions and also started work on a new service that will abstract away and handle checking all external dependencies (ViGEmBus, HidHide present, etc.).

**Reply by nefarius** [reply]
## ViGEmBus not found

![image](https://user-images.githubusercontent.com/286631/149666618-89f494cf-dea2-4885-8dfc-e1ebc9202638.png)

**Reply by nefarius** [reply]
## HidHide not found

And another one 😉 

![image](https://user-images.githubusercontent.com/286631/149673002-0700af9a-bc39-4988-9638-9c8d5a11782e.png)

**Comment by nefarius** [comment]
# 2022-01-17

The Log Tab has been removed. New behaviour will be to only use UI notifications for really important events and instrument more verbose logging to file. This is easier to submit support requests in the future and takes away a lot of verbose spaghetti code that currently writes to the UI from various different classes.

A new "dashboard"-like tab will be introduced which will show the user essential component health in one simple summary (e.g. emulation driver is missing and what to do, settings seem corrupted, performance issue detected, update is available etc.).

**Comment by nefarius** [comment]
# 2022-01-18

Started migrating the "new stuff" to dedicated class library `DS4Windows.Shared.Core`

**Comment by nefarius** [comment]
# 2022-01-22

Continued to rip apart the project into multiple class libraries grouped by core types, application settings, profile settings, device enumeration and more to come.

Also removed deprecated OpenTracing with OpenTelemetry. For now only the Tracing component of OpenTelemetry is gonna be used, Serilog does a fantastic job and will stay until there's good reasons not to. Not a lot is instrumented yet with tracing but the foundation is built.

**Reply by nefarius** [reply]
Now that OpenTelemetry is set up again and works nicely with Jaeger we can start instrumenting some common paths that currently take their time to get spun up, like application boot:

![image](https://user-images.githubusercontent.com/286631/150655247-e109975a-fe0f-4eab-bef8-971432829906.png)

Here we can see, that main window creation and display takes its fair time, so this should be further instrumented and looked into.

**Reply by nefarius** [reply]
Getting to the bottom of this 🔍 

![image](https://user-images.githubusercontent.com/286631/150655964-cd484410-debd-4fc4-ac43-1944f35145bc.png)

**Comment by nefarius** [comment]
# 2022-02-20

Currently very busy with other tasks so not much progress **but** I've commissioned and donated a new app icon, social banner and splash screen to refresh the identity of the app slightly, check it out in the README! 🥳 

![Splashscreen2](https://user-images.githubusercontent.com/286631/154842474-ecd89ecc-9b06-408f-869a-e9cfbe7269fb.png)

See you soon!

**Comment by nefarius** [comment]
# 2022-03-09

Long time no see! I am please to announce that a new contributor found the way into our little corner of the Internet: @Torinth 🥳 

Due to the amount of moving and in-progress code it's been decided that the entire UI system is getting a rewrite/redesign in parallel:

![image](https://user-images.githubusercontent.com/286631/157523890-36f731b8-6c94-475f-bea9-661b7cdf3a9b.png)

This is of course just a mock-up but coding from scratch without the obstacles of the past is sometimes the better choice!

**Comment by nefarius** [comment]
# 2022-03-13

Added [Dependabot](https://github.com/dependabot) configuration so 3rd party packages get checked for updates each week automatically. Working on more components of the input device pipeline.

**Comment by nefarius** [comment]
# 2022-03-22

I am currently buried with other tasks **but** thanks to @Torinth the new UI/UX is getting rapid attention and progress 👏

**Comment by nefarius** [comment]
# 2022-04-23

Things are getting into shape, we have a working prototype that can emulate a controller and managed to shave off one or two more milliseconds of processing speed! More to come soon!

**Comment by nefarius** [comment]
# 2022-04-27

We've got some exciting new experimental approaches to the whole wonky device hiding mechanism in the pipeline, more to come soon 👀

**Comment by nefarius** [comment]
# 2022-05-05

We're experimenting with a new, exciting and promising method to further simplify and automate device hiding without all the pitfalls learned from previous solutions (HidGuardian, HidHide), I'll add some screenshots when I'm ready to show something 👀

**Comment by nefarius** [comment]
# 2022-05-24

Most work currently happens on shaping the new UI and creating an installer. New design uses a Windows Service for tasks requiring elevated permissions, a tray app that is decoupled from the main executable and last but not least a new installer!

Meanwhile a new companion tool takes shape 👀 

![image](https://user-images.githubusercontent.com/286631/169997862-f899df44-461a-4192-8e02-5f048d223255.png)

**Comment by nefarius** [comment]
# 2022-07-04

[Said tool has been published](https://github.com/nefarius/Identinator) 😎

**Comment by nefarius** [comment]
# 2022-10-06

We're still alive and well! Summer is over though, sadly, so as we welcome the dreary cold days, we will fire up the engines in the development dungeon once more and intergrate new features 😁 

Meanwhile for the biggest piece of news: since we couldn't decide on a new name yet, a compromise has been reached!

**Codename: Vapour Input**

This has to do until "public" app branding becomes relevant. But now at least this measure can distinguish us from the currently actively maintained fork(s) of DS4Windows!

Cheers

**Comment by nefarius** [comment]
# 2022-10-16

Huge success achieved today, our new device hiding feature is working; hide and unhide controllers during runtime in a matter of milliseconds! 🥳 

![image](https://user-images.githubusercontent.com/286631/196049273-0961e49e-6939-4bef-af9e-ba3169e270c6.png)

More to come soon 😃

**Comment by nefarius** [comment]
# 2022-11-25

Still alive! 😁 Small update, we successfully migrated the solution to latest-greatest .NET 7! We also managed to shave off 5 megabytes of the resulting binaries 💪 

I'd guess it's safe to assume, that not many newsworthy changes will happen in 2022 so fingers crossed that we may start 2023 refreshed and eager to continue the work 😎 

Cya!

**Comment by nefarius** [comment]
# 2023-01-05

Happy New Year! 🥳 

Let me bring you up to speed what's going on:

- We're currently working on a new exciting method to reliably hide controllers **connected via Bluetooth**, as the system we built for USB is pretty much done and appears to be fast and reliable. The relevant new code will be provided by [yet another class library](https://github.com/nefarius/Nefarius.Utilities.Bluetooth) you can explore and use yourself!
- We migrated and finished support for the Sony DS4 Wireless USB Dongle in both native and hidden/filtered operation. Also added support for a weird DS4 **Rev2** aftermarket controller that for some odd reason spoofs the **Rev1** hardware ID... *Sigh* manufacturers, when will they stop toying with us 😅 
- Support for the Steam Deck/Steam Controller is in the works!
- The main app controller tab got a new look with tiles and displays the emulation mode now!
- Per-launched-game configuration auto-switching support!
- Controllers can be configured to not be filtered/hidden and to not have an emulated output device attached.

Cheers!

**Comment by nefarius** [comment]
# 2023-02-04

Ahoy there! Things are yet again happening at a stellar pace, time to bring you up to speed:

- We've added support for a wide variety of modern **Xbox One Controller variants** as supported input devices!
- Experimental **Bluetooth device hiding** has been added and is almost universally working great!
- The **Steam Deck controller** has been added as a supported input device!
- The **Nintendo JoyCons** have been added as a supported input device!
- Discovery for UWP games has been added!
- Basic **Custom Actions** (like Quick Disconnect via hotkeys) have been introduced!
- Working on the output pipeline (rumble, LED state changes, et.)
- ...and last but not least UI advancements!

Cheers!

**Comment by nefarius** [comment]
# 2023-09-25

Hey there 👋 Been [a bit busy with things](https://docs.nefarius.at/projects/ViGEm/End-of-Life/) but still here 😅 

Cheers!

**Comment by nefarius** [comment]
# 2024-01-17

Belated Happy New Year! 🎉 

Pace is slow, however, changes happened:

- Solution has been upgraded to **.NET 8**
- Various UI components have been migrated to use Source Generators

Cheers!

**Reply by golimonkey** [reply]
Ty for your effort. Cant wait to try it when you release build to the public. DS4Windows is magic, I use it every day. It has some bugs and I cant update to newer versions because of some new bugs (I use fake DS4-s) but I'm hopeing your fork might work better.

**Reply by Hellboy00000** [reply]
Thanks for all your work. 
What happens with Identinator?

**Reply by nefarius** [reply]
Nothing really, it was a proof-of-concept but as time marches on we found more elegant solutions so there isn't really a need to keep it alive or develop it further.

**Comment by God-damnit-all** [comment]
Mister Nefarius, I'm forced to conclude that you don't love us anymore.
